src/key-backup.ts

```diff
@@ -0,0 +1,188 @@
+import JWT from '@tsndr/cloudflare-worker-jwt';
+import { Bindings } from './bindings';
+import { u8ToB64 } from './utils/base64';
+import { R2Bucket } from '@cloudflare/workers-types/2023-07-01';
+import { Context } from './context';
+import { WshimLogger } from './context/logging';
+
+// Cache for the Google Access Token to avoid re-authenticating on every request.
+// Tokens are valid for 1 hour (3600 seconds).
+let cachedAccessToken: { token: string; expiry: number } | null = null;
+
+type ServiceAccountKey = {
+	type: string;
+	project_id: string;
+	private_key_id: string;
+	private_key: string;
+	client_email: string;
+	client_id: string;
+	auth_uri: string;
+	token_uri: string;
+	auth_provider_x509_cert_url: string;
+	client_x509_cert_url: string;
+	universe_domain: string;
+};
+
+const getGoogleAccessToken = async (
+	logger: WshimLogger,
+	serviceAccountKey: string
+): Promise<string> => {
+	if (cachedAccessToken && Date.now() < cachedAccessToken.expiry) {
+		return cachedAccessToken.token;
+	}
+
+	try {
+		const serviceAccount: ServiceAccountKey = JSON.parse(serviceAccountKey);
+		const now = Math.floor(Date.now() / 1000);
+		const expires = now + 3600; // Token valid for 1 hour (Google's max for JWT)
+
+		// Define the JWT payload
+		const payload = {
+			iss: serviceAccount.client_email,
+			scope: 'https://www.googleapis.com/auth/devstorage.full_control',
+			aud: 'https://oauth2.googleapis.com/token',
+			exp: expires,
+			iat: now,
+		};
+
+		const jwt = await JWT.sign(payload, serviceAccount.private_key, { algorithm: 'RS256' });
+
+		const response = await fetch('https://oauth2.googleapis.com/token', {
+			redirect: 'manual',
+			method: 'POST',
+			headers: {
+				'Content-Type': 'application/x-www-form-urlencoded',
+			},
+			body: new URLSearchParams({
+				grant_type: 'urn:ietf:params:oauth:grant-type:jwt-bearer',
+				assertion: jwt,
+			}).toString(),
+		});
+
+		if (!response.ok) {
+			const errorText = await response.text();
+			logger.error('Google OAuth Token Error:', errorText);
+			throw new Error(
+				`Failed to get Google Access Token: ${response.status} ${response.statusText} - ${errorText}`
+			);
+		}
+
+		const data: { access_token: string; expires_in: number } = await response.json();
+
+		// Cache the new token with a slight buffer before actual expiry
```

Comment:
I think this is only cached per worker isolate. Can/should we cache in the colo cache or perhaps with KV?

Reply:
These keys expire quite quickly. To be honest I think caching it even in the same isolate doesn't make much sense.